src/switch.js

```diff
@@ -0,0 +1,271 @@
+   * @param {!Array<!Blockly.utils.Rect>} savedPositions List of rectangles that
+   *     are already on the workspace.
+   */
+  position(metrics, savedPositions) {
+    // Not yet initialized.
+    if (!this.initialized_) {
+      return;
+    }
+
+    const cornerPosition =
+      Blockly.uiPosition.getCornerOppositeToolbox(this.workspace_, metrics);
+    const startRect = Blockly.uiPosition.getStartPositionRect(
+        cornerPosition, new Blockly.utils.Size(WIDTH, HEIGHT),
+        MARGIN_HORIZONTAL,
+        MARGIN_VERTICAL, metrics, this.workspace_);
+
+    const verticalPosition = cornerPosition.vertical;
+    const bumpDirection = verticalPosition ===
+      Blockly.uiPosition.verticalPosition.TOP ?
+      Blockly.uiPosition.bumpDirection.DOWN :
+      Blockly.uiPosition.bumpDirection.UP;
+    const positionRect = Blockly.uiPosition.bumpPositionRect(
+        startRect, MARGIN_VERTICAL, bumpDirection, savedPositions);
+
+    if (verticalPosition === Blockly.uiPosition.verticalPosition.TOP) {
+      this.MultiSelectGroup_.setAttribute(
+          'transform', 'translate(0, ' + HEIGHT + ')');
+    }
+
+    this.top_ = positionRect.top;
+    this.left_ = positionRect.left;
+    this.svgGroup_.setAttribute(
+        'transform', 'translate(' + this.left_ + ',' + this.top_ + ')');
+  }
+  /**
+   * Create the zoom reset icon and its event handler.
+   * @private
+   */
+  createMultiSelectSvg_() {
+    /* This markup will be generated and added to the .svgGroup_:
+    <g class="blocklyMultiSelect">
+      <image width="32" height="32" xlink:href="media/unselect.svg"></image>
+    </g>
+    */
+    this.MultiSelectGroup_ =
+      Blockly.utils.dom.createSvgElement(Blockly.utils.Svg.G, {
+        'class': 'blocklyMultiSelect',
+      }, this.svgGroup_);
+    const MultiSelectSvg = Blockly.utils.dom.createSvgElement(
+        Blockly.utils.Svg.IMAGE, {
+          'width': WIDTH,
+          'height': HEIGHT,
+        },
+        this.MultiSelectGroup_);
+    MultiSelectSvg.setAttributeNS(
+        Blockly.utils.dom.XLINK_NS, 'xlink:href', UNENABLED_IMG);
+
+    // Attach event listeners.
+    this.onMultiSelectWrapper_ = Blockly.browserEvents.conditionalBind(
+        this.MultiSelectGroup_, 'mousedown', null,
+        this.switchMultiSelect_.bind(this));
+  }
+  /**
+   * Handles a mouse down event on the reset zoom button on the workspace.
+   * @param {!Event} e A mouse down event.
+   * @private
+   */
+  switchMultiSelect_(e) {
+    this.workspace_.markFocused();
+    this.enable = !this.enable;
+    this.plugin_.switchMultiSelect(this.enable);
+    Blockly.Touch.clearTouchIdentifier(); // Don't block future drags.
+    e.stopPropagation(); // Don't start a workspace scroll.
+    e.preventDefault(); // Stop double-clicking from selecting text.
+  }
+
+  /**
+   * Updates the multi select icon.
+   * @param {boolean} enable Whether the multi select is enabled.
+   */
+  updateMultiSelect(enable) {
```

Comment:
I would expect this to be called from `switchMultiSelect_`? Is the image being updated on click?

Reply:
> I would expect this to be called from switchMultiSelect_?
It's called in the `switchMultiSelect` from the class in `index.js`, which is also the `this.plugin_`.

> Is the image being updated on click?
Yes. To avoid confusion, I just updated the name into `updateMultiSelectIcon`

Reply:
Gotcha gotcha. Ok, I think this should be reorganized a bit.

Right now it's hard to reason about the program, because it has a circular dependency. I.E. the current control flow is `MultiSelectControls.switchMultiSelect_` -> `WorkspaceMultiSelect.switchMultiselect` -> `MultiSelectControls.updateMultiSelectIcon`.

My suggestion is:
  1) Have the `MultiSelectControls` take in a callback instead of the `WorkspaceMultiSelect` instance. This will remove the circular dependency.
  2) Have `MultiSelectControls.switchMultiSelect_` update the view of the button directly. This makes the control flow clearer.
  3) Split the logic in `WorkspaceMultiSelect.onKeyUp_` and `WorkspaceMultiSelect.onKeyDown_` into subprocedures so that the call to `updateMultiSelectIcon` won't be triggered by the callback passed to `MultiSelectControls`.

Reply:
Okay, I just changed the call to `WorkspaceMultiSelect.switchMultiselect` into a keyboard event simulation that directly applies to `this.workspace`, this removes the need to access the plugin instance.

> Have `MultiSelectControls.switchMultiSelect_` update the view of the button directly. This makes the control flow clearer.

I guess it will be easier to put button view updation into `WorkspaceMultiSelect.onKeyUp_` and `WorkspaceMultiSelect.onKeyDown_`, as we always want to indicate the current selection mode, no matter it is through the button or through the keyboard.

Reply:
I would not recommend using event simulation to trigger logic, because you are likely to end up with side effects that are not intended. E.g. if someone else registers an onKeyUp_ listener for the workspace, clicking your multiselect button will trigger their event listener.

So I still recommend changing the `MultiselectControls` to take in a callback.

Reply:
Just checked and moved all the logics related into MultiselectControls, thanks for the suggestions.